File: apps/studio/authoring/scripts/shared/feed.lua
-- Feed owner. Chooses the active child script and delegates entry events.
local function child_script(state)
    if state.feed.child == "chat-assistant" then
        return "feed/chat-assistant.lua"
    end
    return "feed/idle-waiting.lua"
end

function transition(state, event)
    if event.type == "conversation.new" then
        state.feed.child = "idle-waiting"
        return delegate("feed/chat-assistant.lua", state, event)
    end
    if event.type == "input.submitted" then
        return delegate(child_script(state), state, event)
    end
    if event.type == "present" then
        return delegate(child_script(state), state, event)
    end
    return delegate("feed/chat-assistant.lua", state, event)
end

File: apps/studio/authoring/scripts/shared/feed.ts
import type { Helpers, PolicyEvent, PolicyResult } from "@inpainter/policy-runtime";
import type { SessionState } from "../../policy/src/types.ts";

function childScript(state: SessionState): string {
  if (state.feed.child === "chat-assistant") {
    return "feed/chat-assistant";
  }
  return "feed/idle-waiting";
}

export function transition(
  rawState: unknown,
  event: PolicyEvent,
  helpers: Helpers,
): PolicyResult {
  const state = rawState as SessionState;
  if (event.type === "conversation.new") {
    state.feed.child = "idle-waiting";
    return helpers.delegate("feed/chat-assistant", state, event);
  }
  if (event.type === "input.submitted") {
    return helpers.delegate(childScript(state), state, event);
  }
  if (event.type === "present") {
    return helpers.delegate(childScript(state), state, event);
  }
  return helpers.delegate("feed/chat-assistant", state, event);
}
